### PR TITLE
Add PlayerConnectingEvent 

### LINKED ISSUE #4300: Event Request: Handle Login
I've looked through quite a few classes and can't find a way to hook in simply for a whitelist mod.

What I'm requesting is a post-auth and pre-login event that provides either the profile or network handler with the ability to set the result to deny or cancel the event.

### PR DESCRIPTION
**Pull request by phit**
fires right after whitelist and ban checks, setting a rejection message
or cancelling it will disconnect the player

based on GH-6158
resolves GH-4300

---

I tried to address the feedback in #6158 but not too familiar with the Forge event system in general and how cancelable is usually handled.
Also the event name is quite broad, any better name ideas?

### CODE REVIEW

**Comment by TheCurle** [comment]
How about PlayerConnectedEvent? ServerConnectEvent insinuates the server itself is connecting to something, like another server.

**Comment by phit** [comment]
"connected" sounds like it's already done, but since the whole point of this event is the ability to stop the connecting early I named it "PlayerConnectingEvent" for now

File: src/main/java/net/minecraftforge/event/PlayerConnectingEvent.java
Comment (phit) [inline src/main/java/net/minecraftforge/event/PlayerConnectingEvent.java]: rename here before merge

File: patches/minecraft/net/minecraft/network/login/ServerLoginNetHandler.java.patch
Comment (diesieben07) [inline patches/minecraft/net/minecraft/network/login/ServerLoginNetHandler.java.patch]: Unnecessary whitespace changes.

File: src/main/java/net/minecraftforge/event/PlayerConnectingEvent.java
Comment (diesieben07) [inline src/main/java/net/minecraftforge/event/PlayerConnectingEvent.java:98]: One thing that was [noted in a previous PR](https://github.com/MinecraftForge/MinecraftForge/pull/6158#issuecomment-531923415) is that people should probably not be able to turn an already rejected login into an accepted one.
I agree with that. I might even go so far as to say to not even fire the event if vanilla rejects the player already: Do we really want mods to be able to bypass the vanilla bans and whitelist? Probably not.

**Review by diesieben07 — CHANGES_REQUESTED** [review]
The name `PlayerConnectingEvent` seems fine to me.

File: src/main/java/net/minecraftforge/event/PlayerConnectingEvent.java
Comment (TheSilkMiner) [inline src/main/java/net/minecraftforge/event/PlayerConnectingEvent.java]: Why's this a Fully Qualified Name?

Comment (phit) [reply src/main/java/net/minecraftforge/event/PlayerConnectingEvent.java]: copy pasta :D https://github.com/MinecraftForge/MinecraftForge/blob/1.15.x/src/main/java/net/minecraftforge/event/ServerChatEvent.java#L46

File: src/main/java/net/minecraftforge/event/PlayerConnectingEvent.java
Comment (phit) [reply src/main/java/net/minecraftforge/event/PlayerConnectingEvent.java:98]: Being able to modify the vanilla whitelist and bans is the main point why I made this PR, say you have a mod that adds additional whitelist.jsons from various sources, if the player isn't in the vanilla whitelist the mod wouldn't be able to do that.

Comment (diesieben07) [reply src/main/java/net/minecraftforge/event/PlayerConnectingEvent.java:98]: The vanilla whitelist should just be disabled in that case.

Comment (dries007) [reply src/main/java/net/minecraftforge/event/PlayerConnectingEvent.java:98]: Idealy the event would let you know if the vanilla action was to block and let you overwrite that.

> The vanilla whitelist should just be disabled in that case.

I get many, many support requests for my mod (MCLink) from people who enable the whitelist and expect my mod to then overwrite that setting. I could overwrite that option on every startup, overwrite the command, and do a whole bunch of other, non-user friendly things. But I'd prefer not to do that and just have the event always fire and let the modder handle it.

Comment (diesieben07) [reply src/main/java/net/minecraftforge/event/PlayerConnectingEvent.java:98]: I don't like the idea of mods being able to bypass vanilla whitelist decisions.
If as a user I enable the whitelist I expect it to be enforced. If I don't want it enabled, I don't enable it.

Comment (dries007) [reply src/main/java/net/minecraftforge/event/PlayerConnectingEvent.java:98]: You are not an average user though. An average user expects to install a "whitelist mod" on their server and everything to Just Work™. Explaining to them that they have to install a "whitelisting mod" and then _disable_ the whitelist results in confusion.

Then you also get the added bonus of the following scenario (has happend, I've received complaints about it):
- User has a whitelisted server
- User installs whitelist mod to do some automatic whitelist managent
- User has to disalbe the whitelist because of reasons
- Everything works for a wile
- User decides to remove the mod for any reason
- Server is now no longer protected by any whitelist because they had to disable it previously

I agree that this has potential to be abused. But so do many, many other things. And when they do get abused, people stop using the mod or they complain.
I'd say make it a "paranoid whitelist" config option, but honestly, that's gonna be even more confusing.

Comment (DeliciousJaffa) [reply src/main/java/net/minecraftforge/event/PlayerConnectingEvent.java:98]: Having a mod-specific "lockout whitelist" is confusing to users by itself.
I'd expect to be able to enable the vanilla whitelist when maintenance on a server is required and not have to search through (potentially multiple) mod config files.
A mod might not even implement a method to have a "lockout" resulting in a user having to disable mods and restart a server to achieve a basic task which would be otherwise accomplished by typing `whitelist on`.

**Comment by diesieben07** [comment]
After further discussion on Discord:

- This will _definitely_ not go in it's current form (mods should not be able to overrule vanilla whitelist, etc.).
- It is questionable why this needs to be an event at all. Why do multiple mods need to control the whitelist? Why do mods need to control the whitelist at all? Your use-case of MCLink (the only one provided here) is better solved by e.g. a cronjob which sets the whitelist.

Please provide any other use-cases you might have for this.

**Comment by phit** [comment]
- first point is fine with me, I'll change it so it only fires if the player is allowed to join by vanilla
- multiple mods don't really need to control it, I suppose you are suggesting implementing this like permissions with one handler? cronjobs are useless it means you need control the underlying system, which isn't the case for most servers with server providers, and even if you do, you still need to somehow trigger the reload of the whitelist via e.g. RCON, having it all handled in a mod is a lot cleaner
here's an example mod I've used in the past to achieve similar results:
https://www.curseforge.com/minecraft/mc-mods/whitelister there's also quite a few examples on CF https://www.curseforge.com/minecraft/mc-mods/search?category=&search=Whitelist

**Comment by dries007** [comment]
On the first: This is already possible, just not very convenient. I really don't understand why there are any legitimate conserns about this. Bukkit has been doing this for years. A mod can use any trigger to disable the whitelist all together, including a login attempt by someone, so if you wanted to you could already bypass the whitelist. If you cannot trust a mod with this kind of power, why would you trust them with full read/write/execute permissions on your computer? But enough of that.

On the second:
- A cron job relies on a periodic update. MCLink is live. It checks APIs on the fly which is a large part of the appeal. It works the moment you click the subscribe button. You can never have that with a periodic update.
- Most simple server hosts don't have cron jobs.
- People don't want the hassle of configuring cron jobs. They already ask me for simpler things than editing a config file because it's confusing.
- Separation of conserns. With a normal event 1 mod can be used per system, e.g. Discord, Patreon, Twitch, ...
- This works and exists in Bukkit, therefore people expect it to "just work" in Forge as well.

EDIT: Oh, just FYI: Feel free to @-me on discord (@dries007#8913), I'd love to discuss more if you prefer that platform or if you want to know more about MCLink.

**Comment by diesieben07** [comment]
Having it as an event is not trivial.
Who decides on the order? If you have two whitelist mods, the first one rejects the player because they are not on the whitelist... then the 2nd one says "oh no, this guy is on the whitelist" and removes the rejection. But what if the first mod was a banlist mod and the overruling of the rejection makes no sense?

This is not a trivial problem that is not solved by simply firing and event and letting people fight over the contents of the rejection message.

**Comment by dries007** [comment]
The priority system by Forge can be used for ordering, it's not very granular, but it does work.

The event tracks the origianal status (aka what vanilla would have done) so mods can act accordingly:
- A whitelist mod would only turn vanilla "non whitelist" rejections into OK and not touch any event with another initial state.
- A collective ban / reputation mod could be configured to block logins regardless of any previous state.
- A VIP list mod could overwrite the vanilla "full server" state only to allow sertain groups of people to join regardless.

I'd say, leave it up to the modders to decide based on the usecase they want to cover.

Altough it might be benefitial to add an enum that tracks the state, so the originalMessage doesn't have to be parsed/compared by every listener.

**Comment by diesieben07** [comment]
> The priority system by Forge can be used for ordering, it's not very granular, but it does work.

it is not user controllable. What will happen is that mods will put highest priority on everything because their mod is the most important.

As for your other points: Yes, it is possible. But it is not solved by simply creating an event and allowing people to set the rejection message (which is the current state of the PR).

**Comment by Claycorp** [comment]
I have a mod I just recently made that would make use of this also.

Many times I've needed to keep players out of a server while fixing something that has gone wrong while the server is live. I don't want to be constantly telling people to leave or mess with whitelist/server settings to lock out the server that may require a restart for a simple task. Thus the easy option is to prevent people from joining unless they are OP'ed. Easily activated and deactivated all while the server is running and causes the least interference/possible issues.


Also isn't the whole "Who decides on the order" argument moot? Most, if not all, modifiable events have the same issue yet they still exist with the same problem that makes adding this an issue. You can't have a mod that stops someone from doing X and have a mod that allows someone to do X at the same time or even some of them where you can't do two of the same thing at the same time. Users also don't have direct control over what's picked but devs are more than capable of fixing the issue or providing options around it. Otherwise users will find a new mod that does both or just not use one of them.

**Comment by dries007** [comment]
> As for your other points: Yes, it is possible. But it is not solved by simply creating an event and allowing people to set the rejection message (which is the current state of the PR).

I'm willing to put in the effort to modify this PR with the author or create a new one which addresses this, but I first want some consensus on what it should approximately look like to avoid straight up rejection.

**Comment by diesieben07** [comment]
The order argument is not mood. Most events exist not to modify vanilla things per-se but to add a hook to vanilla things _when your mod is involved_.
Take UseHoeEvent: You won't do anything here if the block being target isn't yours. As such: Order is a non-issue.
Here however the express purpose is to modify vanilla things and also do it from multiple mods at once.

@dries007 I'm not the one making the decision whether things go in or not. https://github.com/MinecraftForge/MinecraftForge/pull/6889#issuecomment-666619393 was me relaying to you the opinion of the project maintainer.

**Comment by TheCurle** [comment]
There has been talk of this PR in other platforms, and the consensus is that this needs large architectural changes to work.
Not to mention it's stale. 

As such, I'll close it. If this is something that modders desperately need (it seems not, given that this is possible by other means) then you need to talk with the triage team.

**Pull request closed without merging**
